Add ability to find config files in multiple dirs

veerajsk opened the pull request:
Wrike Task: https://www.wrike.com/open.htm?id=1061561076

- The engine now has a set of directories where it looks for config files recursively
- If the roleset with the given file name is not found (recursive search), then the engine looks for any roleset file in the specified directories but not recursively
- The beh & cnd files are now referred to only by their file name & not by their relative path to the plans folder
- The plan, role & task config keys are no longer required since the search is recursive & hence these are removed

veerajsk committed:
Add ability to find config files in multiple dirs

- The engine now has a set of directories where it looks for config
files recursively
- If the roleset with the given file name is not found (recursive
search), then the engine looks for any roleset file in the specified
directories but not recursively
- The beh & cnd files are now referred to only by their file name &
not by their relative path to the plans folder
- The plan, role & task config keys are no longer required since the
search is recursive & hence these are removed

veerajsk committed:
Merge branch 'devel' into vsk_parse_multiple_folders

veerajsk committed:
Accept multiple config folders in the AlicaContext

veerajsk committed:
Consider only the file name in referenced elements

If the referenced element has a path instead of a name only consider the
name Eg. if a behaviour is referenced by a path (relative or absolute)
to the .beh file then only consider the file name since we perform
a search for the file now instead of directly using the path.

veerajsk committed:
Fix tests

ntfshard commented on alica_engine/include/engine/AlicaContext.h:
```diff
@@ -95,12 +96,12 @@ struct AlicaContextParams
      */
-    AlicaContextParams(const std::string& agentName, const std::string& configPath, const std::string& roleSetName, const std::string& masterPlanName,
-            bool stepEngine = false, const AgentId agentID = InvalidAgentID)
+    AlicaContextParams(const std::string& agentName, const std::unordered_set<std::string>& configPaths, const std::string& roleSetName,
```
maybe we want to have user-defined order instead of unpredictable unordered_set's order

veerajsk replied:
yeah, that would be better, I'll change to std::vector

veerajsk committed:
Fix turtlesim

ntfshard commented on alica_engine/src/engine/AlicaContext.cpp:
```diff
@@ -131,30 +131,48 @@ VariableSyncModule& AlicaContext::editSyncModule()
+                              << badFile.msg << ", will continue the search in the next config directory" << std::endl;
+                }
+            }
+            return node;
```
Return node even if error occurred? Perhaps we should continue search?

veerajsk replied:
yes, I'll remove the return

ntfshard commented on alica_engine/src/engine/AlicaContext.cpp:
```diff
@@ -131,30 +131,48 @@ VariableSyncModule& AlicaContext::editSyncModule()
+                }
+            }
+            return node;
+        }
```
copy-paste

ntfshard commented on supplementary/alica_ros2/alica_ros2_turtlesim/src/base.cpp:
```diff
@@ -33,7 +33,7 @@ Base::Base(rclcpp::Node::SharedPtr nh, rclcpp::Node::SharedPtr priv_nh, const st
     spinner.add_node(priv_nh);
     // Initialize Alica
-    ac = new alica::AlicaContext(AlicaContextParams(name, path + "/etc/", roleset, master_plan, false, agent_id));
+    ac = new alica::AlicaContext(AlicaContextParams(name, {path + "/etc/"}, roleset, master_plan, false, agent_id));
```
consider use smart pointer (several places)

veerajsk replied:
would prefer making unrelated changes in a separate PR & address all places where smart pointers can be used in that PR

ntfshard commented on alica_engine/include/engine/modelmanagement/ModelManager.h:
```diff
@@ -53,9 +50,16 @@ class ModelManager
-    std::string getBasePath(const std::string& configKey);
+    // Recursively look for a file named <fileName> in _domainConfigFolders & return the absolute path to the file. The returned path is empty if the file is
+    // not found
+    std::string findFile(const std::string& fileName);
```
const?

veerajsk replied:
I think it is better to return non const if returning by value for 2 reasons:
1. Returning by const can prevent moving from the returned object which unnecessarily hurts performance eg. ` file = findFile` will always assigns by copy (instead of move) if returned by const
2. The returned value is anyway in the caller's scope & I prefer the caller to have the flexibility to own a non-const version & convert to const if he does really need it.

ntfshard replied:
I meant const qualifier of member function

ntfshard commented on alica_engine/include/engine/modelmanagement/ModelManager.h:
```diff
@@ -53,9 +50,16 @@ class ModelManager
+     * @param dir directory to search in (not recursively)
+     * @return The first default role set it finds.
+     */
+    std::string findDefaultRoleSet();
```
const?

ntfshard replied:
const qualifier of member function

ntfshard commented on alica_engine/src/engine/modelmanagement/ModelManager.cpp:
```diff
@@ -30,69 +30,34 @@
+        }
 
         Logging::logDebug(LOGNAME) << "fileToParse: " << fileToParse;
+        filesParsed.push_back(filesToParse.front());
```
`push_back(fileToParse)`?

veerajsk replied:
fileToParse is the complete path where filesToParse.front() is just the name of the file

ntfshard commented on alica_engine/src/engine/modelmanagement/ModelManager.cpp:
```diff
@@ -120,53 +87,52 @@ Plan* ModelManager::loadPlanTree(const std::string& masterPlanName)
+        std::vector<std::string> files = essentials::FileSystem::findAllFiles(folder, alica::Strings::roleset_extension);
+
+        // find default role set and return the first valid one found
+        for (std::string file : files) {
```
by copy?

veerajsk replied:
I'll change it to reference

ntfshard reviewed (commented):
New code not covered by tests, correct me if I wrong

ntfshard commented:
and in general, I guess we should start using https://en.cppreference.com/w/cpp/filesystem

veerajsk commented:
> and in general, I guess we should start using https://en.cppreference.com/w/cpp/filesystem

I agree. But again, I would defer to another PR in the interest of keeping this PR small & focused

veerajsk commented:
> New code not covered by tests, correct me if I wrong

There is a test for parsing the plan files already & it verifies this change as well for a single folder config path. I am adding another test for multiple folders (PR is still marked draft for this reason)

veerajsk committed:
Review comments: Change unordered_set to vector

For config paths container, use vector instead of unordered_set so that
user order can be maintained. In this case, the search for files will
proceed from index 0 which means that will also be the order of priority
to find files.
Couple of other very minor fixes

veerajsk committed:
Add multi config folder plan parsing unit test

veerajsk committed:
Fix comment

veerajsk committed:
Remove unnecessary include

veerajsk committed:
Minor fix

veerajsk committed:
More fixes

veerajsk committed:
Log fixes

veerajsk committed:
Merge branch 'devel' into vsk_parse_multiple_folders

veerajsk committed:
Merge remote-tracking branch 'origin/devel' into vsk_parse_multiple_folders

veerajsk committed:
Merge branch 'devel' into vsk_parse_multiple_folders

ntfshard commented on alica_engine/src/engine/AlicaContext.cpp:
```diff
@@ -131,30 +131,53 @@ VariableSyncModule& AlicaContext::editSyncModule()
+                return node;
+            } catch (YAML::BadFile& badFile) {
+                if (Logging::isInitialized()) {
+                    Logging::logWarn(LOGNAME) << "Agent wise config file could not be parsed, file: " << configFilePath << ", error details- " << badFile.msg
+                                              << ", will continue the search in the next config directory";
+                } else {
+                    std::cerr << "[WARN] " << LOGNAME << " Agent wise config file could not be parsed, file:  " << configFilePath << ", error details - "
+                              << badFile.msg << ", will continue the search in the next config directory" << std::endl;
```
This is a same error messages, perhapce using variable could help to avoid copy-paste

ntfshard replied:
Same below

Dale-Koenig replied:
IMO maybe logging component itself should output a generic msg if it is used pre-initialization rather than having custom error msgs in multiple places

veerajsk replied:
yes, I can take a look into implementing something like this in the logging component, because we have this issue in other places as well eg. some user code using the logger before it is initialised. I have observed this when other engine components are set before the logger is set

ntfshard commented on alica_engine/src/engine/AlicaContext.cpp:
```diff
@@ -131,30 +131,53 @@ VariableSyncModule& AlicaContext::editSyncModule()
-YAML::Node AlicaContext::initConfig(const std::string& configPath, const std::string& agentName)
+YAML::Node AlicaContext::initConfig(const std::vector<std::string>& configPaths, const std::string& agentName)
 {
     YAML::Node node;
```
redundant variable

ntfshard commented on alica_engine/src/engine/modelmanagement/ModelManager.cpp:
```diff
@@ -120,53 +86,54 @@ Plan* ModelManager::loadPlanTree(const std::string& masterPlanName)
+
+        // find default role set and return the first valid one found
+        for (const auto& file : files) {
+            YAML::Node node;
```
variable scope can be reduced; technically you can use `auto node = YAML::LoadFile(file);`

ntfshard commented on alica_engine/src/engine/modelmanagement/factories/Factory.cpp:
```diff
@@ -45,28 +45,22 @@ int64_t Factory::getReferencedId(const std::string& idString)
+        } else if (std::find(std::begin(modelManager->filesParsed), std::end(modelManager->filesParsed), fileName) == std::end(modelManager->filesParsed) &&
+                   std::find(std::begin(modelManager->filesToParse), std::end(modelManager->filesToParse), fileName) == std::end(modelManager->filesToParse)) {
+            modelManager->filesToParse.push_back(fileName);
         }
```
perhaps else branch also should be here

veerajsk replied:
else is not required, because there is nothing to do in it. Basically the else block will be hit when the file has a valid extension but is already either parsed or included in the list of files to be parsed. Therefore nothing needs to be done in either case.

bjoernschroeder reviewed (commented):
I dont have much to add to the already made review comments. PR is LGTM after resolving them.

Dale-Koenig reviewed (commented):
Please make sure lbc+lbc bridge work without any changes and add backwards compatible (deprecated) functions where necessary

veerajsk committed:
Fix review comments

veerajsk committed:
Check if logging is initialised in the logger

veerajsk committed:
Merge branch 'devel' into vsk_parse_multiple_folders

veerajsk committed:
Fix plan parser test

veerajsk committed:
Bring back single config path constructor as deprecated

ntfshard commented on alica_engine/src/engine/modelmanagement/ModelManager.cpp:
```diff
@@ -30,69 +30,33 @@
     while (filesToParse.size() > 0) {
-        std::string fileToParse = filesToParse.front();
-        filesToParse.pop_front();
+        std::string fileToParse;
+        if (fileToParse = findFile(filesToParse.front()); fileToParse.empty()) {
```
technically you can initialize variable on creation
```suggestion
        const auto fileToParse = findFile(filesToParse.front());
        if (fileToParse.empty()) {
```

ntfshard commented on alica_engine/src/engine/modelmanagement/ModelManager.cpp:
```diff
@@ -120,53 +86,53 @@ Plan* ModelManager::loadPlanTree(const std::string& masterPlanName)
 RoleSet* ModelManager::loadRoleSet(const std::string& roleSetName)
 {
     std::string roleSetPath;
```
```suggestion
    auto roleSetPath = findFile(roleSetName + alica::Strings::roleset_extension);
```
and you don't need such long expression inside `if`

ntfshard commented on supplementary/alica_ros1/alica_tests/src/test/test_alica_engine_plan_parser.cpp:
```diff
@@ -165,418 +165,446 @@ class AlicaEngineTest : public AlicaTestFixture
-                    break;
+                cout << endl;
+                break;
+            case 1402488893641:
```
can't review this changes properly

veerajsk replied:
I didn't make any changes to this code. The only thing I did is to move the plan parsing checks into a function in the base class & call it in both tests

ntfshard reviewed (commented):
I guess it's ok

veerajsk committed:
Update alica_engine/src/engine/modelmanagement/ModelManager.cpp

Co-authored-by: Maksim Derbasov <ntfs.hard@gmail.com>

veerajsk committed:
Review comment

veerajsk commented:
> I guess it's ok

I addressed all the review comments, so could you please approve if it's ok?

ntfshard commented on alica_engine/src/engine/modelmanagement/ModelManager.cpp:
```diff
@@ -30,69 +30,33 @@
-
 Plan* ModelManager::loadPlanTree(const std::string& masterPlanName)
 {
     std::string masterPlanPath;
```
same case here, if you want

veerajsk replied:
I think this is nit & we need this change to unblock changes in the backend. So merging
